env/env.cc

```diff
@@ -476,6 +476,13 @@ const std::shared_ptr<FileSystem>& Env::GetFileSystem() const {
 std::unique_ptr<Env> NewCompositeEnv(std::shared_ptr<FileSystem> fs) {
   return std::unique_ptr<Env>(new CompositeEnvWrapper(Env::Default(), fs));
 }
+
+std::unique_ptr<Env> NewCompositeEnv(std::shared_ptr<FileSystem> fs,
+                                     std::shared_ptr<Env>* shared_env) {
+  assert(shared_env != nullptr);
+  if (!*shared_env) shared_env->reset(Env::Default());
```

Comment:
Won't this cause `Env::Default()` to be double deleted (once when the last `shared_ptr` is destroyed, again when the static `Env` it points to goes out of scope?

Reply:
The windows implementation does this (port/win/env_default.cc) : `std::call_once(winenv_once_flag, []() { envptr = new WinEnv(); });`

So it should work as long as Env::Default() is only used for NewComposite envs, right? 

Maybe a better solution would be to add a new Env:Default() that returns a shared pointer, avoiding the out-of-scope-racing entirely?

Thanks for reviewing!

Reply:
> So it should work as long as Env::Default() is only used for NewComposite envs, right?

For Windows yes, but not for Linux as its `Env::Default()` uses a function-local static: https://github.com/facebook/rocksdb/blob/722ebba834d550ec03d481ecbb7c1e7857cb9fb7/env/env_posix.cc#L508-L509.

Can we call the `NewCompositeEnv(std::shared_ptr<FileSystem> fs)` overload in case of `!*shared_env`? It seems we do not need a `shared_ptr` to guard the `CompositeEnvWrapper`'s `Env` when it is `Env::Default()`.

Reply:
I've now removed the new NewCompositeEnv overload, and added a getter in stead. This allows for composite envs to share thread pools and for the base env to join those threads before the composite envs are destroyed(avoiding the race) - without destroying Env::Default(). Does this make sense?

Reply:
Ping @ajkr , could you have a look at the new revision of the branch? Looks neater to me.  Cheers!